runtime/jvmti/suspendhelper.cpp

```diff
@@ -35,33 +35,48 @@ suspendThread(J9VMThread *currentThread, jthread thread, UDATA allowNull, UDATA
 	if (rc == JVMTI_ERROR_NONE) {
-		if (targetThread->publicFlags & J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND) {
-			rc = JVMTI_ERROR_THREAD_SUSPENDED;
-		} else {
-			if (targetThread->publicFlags & J9_PUBLIC_FLAGS_STOPPED) {
-				/* Do not suspend dead threads. Mirrors SUN behaviour. */
-				rc = JVMTI_ERROR_THREAD_NOT_ALIVE;
+#if JAVA_SPEC_VERSION >= 19
+		j9object_t threadObject = (NULL == thread) ? currentThread->threadObject : J9_JNI_UNWRAP_REFERENCE(thread);
+		if (NULL != targetThread)
+#endif /* JAVA_SPEC_VERSION >= 19 */
+		{
+			if (OMR_ARE_ANY_BITS_SET(targetThread->publicFlags, J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND)) {
+				rc = JVMTI_ERROR_THREAD_SUSPENDED;
 			} else {
-				if (currentThread == targetThread) {
-					*currentThreadSuspended = TRUE;
+				if (OMR_ARE_ANY_BITS_SET(targetThread->publicFlags, J9_PUBLIC_FLAGS_STOPPED)) {
+					/* Do not suspend dead threads. Mirrors SUN behaviour. */
+					rc = JVMTI_ERROR_THREAD_NOT_ALIVE;
 				} else {
-					currentThread->javaVM->internalVMFunctions->internalExitVMToJNI(currentThread);
-					omrthread_monitor_enter(targetThread->publicFlagsMutex);
-					VM_VMAccess::setHaltFlagForVMAccessRelease(targetThread, J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND);
-					if (VM_VMAccess::mustWaitForVMAccessRelease(targetThread)) {
-						while (J9_ARE_ALL_BITS_SET(targetThread->publicFlags, J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND | J9_PUBLIC_FLAGS_VM_ACCESS)) {
-							omrthread_monitor_wait(targetThread->publicFlagsMutex);
+					if (currentThread == targetThread) {
+						*currentThreadSuspended = TRUE;
+					} else {
+						currentThread->javaVM->internalVMFunctions->internalExitVMToJNI(currentThread);
+						omrthread_monitor_enter(targetThread->publicFlagsMutex);
+						VM_VMAccess::setHaltFlagForVMAccessRelease(targetThread, J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND);
+						if (VM_VMAccess::mustWaitForVMAccessRelease(targetThread)) {
+							while (J9_ARE_ALL_BITS_SET(targetThread->publicFlags, J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND | J9_PUBLIC_FLAGS_VM_ACCESS)) {
+								omrthread_monitor_wait(targetThread->publicFlagsMutex);
+							}
 						}
+						omrthread_monitor_exit(targetThread->publicFlagsMutex);
+						currentThread->javaVM->internalVMFunctions->internalEnterVMFromJNI(currentThread);
 					}
-					omrthread_monitor_exit(targetThread->publicFlagsMutex);
-					currentThread->javaVM->internalVMFunctions->internalEnterVMFromJNI(currentThread);
+					Trc_JVMTI_threadSuspended(targetThread);
 				}
-				Trc_JVMTI_threadSuspended(targetThread);
 			}
 		}
+#if JAVA_SPEC_VERSION >= 19
+		else if (IS_VIRTUAL_THREAD(currentThread, threadObject)) {
+			jint vthreadState = J9VMJAVALANGVIRTUALTHREAD_STATE(currentThread, threadObject);
+			if (OMR_ARE_ANY_BITS_SET(vthreadState, JVMTI_VTHREAD_STATE_SUSPENDED)) {
+				rc = JVMTI_ERROR_THREAD_SUSPENDED;
+			} else {
+				J9VMJAVALANGVIRTUALTHREAD_SET_STATE(currentThread, threadObject, vthreadState | JVMTI_VTHREAD_STATE_SUSPENDED);
+			}
+		}
```

Comment:
Error case here for non-virtual threads?

Reply:
Same as https://github.com/eclipse-openj9/openj9/pull/15903#discussion_r976521831.